flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandler.java

```diff
@@ -0,0 +1,102 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.runtime.checkpoint.CheckpointException;
+import org.apache.flink.runtime.io.network.api.CancelCheckpointMarker;
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+import org.apache.flink.runtime.io.network.buffer.BufferReceivedListener;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static java.lang.String.format;
+import static org.apache.flink.runtime.checkpoint.CheckpointFailureReason.CHECKPOINT_DECLINED_SUBSUMED;
+
+class AlternatingCheckpointBarrierHandler extends CheckpointBarrierHandler {
+	private final CheckpointBarrierAligner alignedHandler;
+	private final CheckpointBarrierUnaligner unalignedHandler;
+	private CheckpointBarrierHandler activeHandler;
+
+	AlternatingCheckpointBarrierHandler(CheckpointBarrierAligner alignedHandler, CheckpointBarrierUnaligner unalignedHandler, AbstractInvokable invokable) {
+		super(invokable);
+		this.activeHandler = this.alignedHandler = alignedHandler;
+		this.unalignedHandler = unalignedHandler;
+	}
+
+	@Override
+	public void releaseBlocksAndResetBarriers() {
+		activeHandler.releaseBlocksAndResetBarriers();
+	}
+
+	@Override
+	public boolean isBlocked(int channelIndex) {
+		return activeHandler.isBlocked(channelIndex);
+	}
+
+	@Override
+	public void processBarrier(CheckpointBarrier receivedBarrier, int channelIndex) throws Exception {
+		CheckpointBarrierHandler previousHandler = activeHandler;
+		activeHandler = receivedBarrier.isCheckpoint() ? unalignedHandler : alignedHandler;
+		abortPreviousIfNeeded(receivedBarrier, previousHandler);
+		activeHandler.processBarrier(receivedBarrier, channelIndex);
+	}
+
+	private void abortPreviousIfNeeded(CheckpointBarrier barrier, CheckpointBarrierHandler prevHandler) throws IOException {
+		if (prevHandler != activeHandler && prevHandler.isCheckpointPending() && prevHandler.getLatestCheckpointId() < barrier.getId()) {
+			prevHandler.releaseBlocksAndResetBarriers();
+			notifyAbort(
+				prevHandler.getLatestCheckpointId(),
+				new CheckpointException(
+					format("checkpoint %d subsumed by %d", prevHandler.getLatestCheckpointId(), barrier.getId()),
+					CHECKPOINT_DECLINED_SUBSUMED));
+		}
+	}
+
+	@Override
+	public void processCancellationBarrier(CancelCheckpointMarker cancelBarrier) throws Exception {
+		activeHandler.processCancellationBarrier(cancelBarrier);
+	}
+
+	@Override
+	public void processEndOfPartition() throws Exception {
+		alignedHandler.processEndOfPartition();
```

Comment:
A bit confuse here why we need call both handlers, should only the active handler work at the same time?

Reply:
I this case both handlers need to update their number of open input channels.